Add unit_value to Resource Objects and Children

dean-krueger opened the pull request:
# Summary of Changes

A member variable "unit_value_" has been added to all resource-type objects (`Resource`, `Product`, and `Material`) and appropriate functions (`Absorb`, `Clone`, `Extract`) have been updated to have reasonable behavior concerning the new `unit_value_` member.

# Related CEPs and Issues

This PR is related to:

- Closes #1885 

# Associated Developers

# Design Notes

Wasn't sure how to handle packaging. May need to discuss in person, or here, either works.

# Testing and Validation

Built and tested Cyclus on my local machine. Added unit tests. Clean Build Cyclus and Cycamore, all tests for both passing.

# Checklist

 - [x] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide.
 - [x]  Compile and run locally.
 - [x]  Add or update tests.
 - [x]  Document if needed.
 - [x]  Follow style guidelines.
 - [x]  Update the changelog.
 - [x]   Run clang-format
 - [ ]  Address all review comments.
Reviewers, please refer to the Cyclus [Guide for Reviewers](https://fuelcycle.org/kernel/pr_review.html).

dean-krueger committed:
added unit_value_ to resource and updated absorb functions in material and product to update unit_value_ with the average

dean-krueger committed:
added tests, tests revealed bugs, bugs squashed, tests re-ran successfully

dean-krueger commented:
Need to do changelog and clang-format, but need to run to something real quick, so I'll do those later today.

dean-krueger committed:
changelog and clang-format

gonuke commented on src/resource.h:
```diff
@@ -112,6 +119,8 @@ class Resource {
   static int nextstate_id_;
   static int nextobj_id_;
   int state_id_;
+  // Unit value of the resource.
+  double unit_value_;
```
(Pedantic) let's put this either after or before all the ID variables.

dean-krueger replied:
Oops, thought I did, but I guess I missed the `int obj_id_` below. My bad!

gonuke commented on src/product.cc:
```diff
@@ -78,6 +86,7 @@ std::string Product::package_name() {
   return package_name_;
 }
 
+// Not sure what to do here with the unit value and packaging...
```
This should behave the same way as `Extract` from the point if view of the unit value

gonuke replied:
But I think I'll need to remember why this extract needs to be special and add better comments for why

gonuke commented on src/material.cc:
```diff
@@ -86,7 +89,8 @@ Material::Ptr Material::ExtractComp(double qty, Composition::Ptr c,
 
   qty_ -= qty;
-  Material::Ptr other(new Material(ctx_, qty, c, Package::unpackaged_name()));
+  Material::Ptr other(
```
Need to add similar to `PackageExtract`

dean-krueger replied:
Done. Good catch.

gonuke commented on tests/resource_tests.cc:
```diff
@@ -72,17 +76,29 @@ TEST_F(ResourceTest, MaterialAbsorbGraphid) {
 TEST_F(ResourceTest, MaterialExtractTrackid) {
   int obj_id = m1->obj_id();
-  Material::Ptr m3 = m1->ExtractQty(2);
+  Material::Ptr m4 = m1->ExtractQty(2);
```
I wonder if we should rename all the local objects to avoid potential name collisions in the future
```suggestion
  Material::Ptr lm1 = m1->ExtractQty(2);
```
Add 'l' for "local" and restart numbering at 1?

gonuke reviewed (changes requested):
This looks like a good start. In my next review, I'll more carefully look for places where changes should exist by may be forgotten.

dean-krueger committed:
added UnitValue() calls to PackageExtract to give the new resource they create a value, some variable reordering

dean-krueger committed:
changed m4 to lm1, p4 to lp1, etc in resource tests

dean-krueger commented:
So in my search I found the following method in `res_tracker.cc`:

```
void ResTracker::Record(bool bumpId) {
  if (bumpId) {
    res_->BumpStateId();
  }
  ctx_->NewDatum("Resources")
      ->AddVal("ResourceId", res_->state_id())
      ->AddVal("ObjId", res_->obj_id())
      ->AddVal("Type", res_->type())
      ->AddVal("TimeCreated", ctx_->time())
      ->AddVal("Quantity", res_->quantity())
      ->AddVal("Units", res_->units())
      ->AddVal("QualId", res_->qual_id())
      ->AddVal("PackageName", res_->package_name())
      ->AddVal("Parent1", parent1_)
      ->AddVal("Parent2", parent2_)
      ->Record();
  res_->Record(ctx_);
}
```

I'm not totally sure if this is the one you were talking about with the `Tracker` bit of what you were saying, but that __appears__ to be the only place in the code where the `Resources` table gets specifically written to (that method gets called when resources are created, modified, split, combined, or repackaged, though, through the `tracker_` object methods). It LOOKS like this table is the one that records all the other properties of a resource (id, type, units, etc), so to me it seems like this is where we would want to write out its value (especially since this method gets called whenever resources are modified, too.

Interested to hear your thoughts!

gonuke commented:
> So in my search I found the following method in `res_tracker.cc`:

Good find!  I think that's the place we want to do it, for sure.

dean-krueger commented:
On it!

dean-krueger committed:
Added UnitValue column to the Resources table

gonuke commented on src/material.cc:
```diff
@@ -15,14 +15,17 @@ const ResourceType Material::kType = "Material";
   return m;
 }
 
 Material::Ptr Material::CreateUntracked(double quantity, Composition::Ptr c) {
```
Should we allow untracked materials to be given a `unit_value` upon creation?  One can always call `SetUnitValue()` after the fact, but it might be more convenient if we allow it directly during creation?

gonuke reviewed (changes requested):
One last comment and then I think I'm ready to move forward and address anything else as it comes.

dean-krueger committed:
added ability to give untracked materials a unit_value, set default of this unit_value to 0.0

dean-krueger committed:
added a constant for unset unit values called kUnsetUnitValue and set it to quiet_nan such that it won't break anything, but carry through in calculations as nan so it's unambiguous

dean-krueger committed:
fixed unit tests, clang formatted changes

gonuke reviewed (approved):
I think this looks good now.  It will be hard to exhaustively explore all the places that this will matter, so in the future, we'll need to carefully look for places where the unit value gets modified or reset to default accidentally.
